[WIP] Flexible Byte Layout

mikeal opened the pull request:
I got the new `js-unixfsv2` working using `ipld-schema-gen`. It’s awesome, but now my brain is mush and I need a rest.

I wanted to get this posted before I crash though. The structure I landed on for the layout of the byte data is very interesting and general purpose.

I figured out a way, using recursive unions, to create large DAG’s without encoding the layout algorithm into the data structure. This means that we can create a general read implementation for **any** layout we decide to implement in the future. 

All you do is implement `read()` methods for 3 advanced layouts and those readers essentially call each other all the way through the DAG. So you can mix and match all the basic layout components on the write() side to your heart’s content. Since it’s a union, we also wouldn’t have trouble adding another one in the future, but the problem of “representing a linear list of bytes” isn’t so large that I anticipate needing another one.

mikeal commented:
I was so burned out last night I forgot to reference the current implementation https://github.com/ipld/js-unixfsv2/blob/master/src/data.js

This is using a slightly different schema because `rootType` is not yet implemented in the js schema parser, so my `schema-gen` library has a hack that links an advanced layout to a schema by using the same name. This will all be updated and put in-line with this schema once `rootType` is fully implemented.

rvagg commented on schema-layer/data-structures/Data.md:
```diff
@@ -0,0 +1,62 @@
+  | Bytes "bytes"
+  | &Bytes "bytesLink"
+  | ByteLinks "byteLinks"
+  | NestedByteList "nbl"
```
Oh the bikeshed potential of this section!  You obviously have a "that's too long" measure in your head that made you shorten `"nestedByteList"` to `"nbl"` but unless you can be clear about how to measure that then you're going to have to defend the long-form of the other 3 from people who instinctively want to shave bytes.

I'd argue for more consistency whichever way you go. Easier to draw a firm line and defend choices.

rvagg replied:
would still like discussion on this

rvagg commented on schema-layer/data-structures/Data.md:
```diff
@@ -0,0 +1,62 @@
+
+advanced ByteLinksLayoutAdvanced {
+  rootType ByteLinksLayout
+ }
```
leading space

mikeal commented:
> names of things

I struggled a lot with the naming, not just the union key names but also pluralization. I must have changed some of these names 2 or 3 times throughout the coarse of development.

As a result, I’m going to be a real push over on this, I’m basically willing to change the naming convention to whatever the first person says, because I’m not happy with my existing choices here and am happy to defer to the first person I hear ;)

mikeal commented:
We need to solidify this and get it finished.

@warpfork @rvagg any pressing concerns or final changes you’d like to see before it gets merged.

warpfork commented on schema-layer/data-structures/Data.md:
````diff
@@ -0,0 +1,62 @@
+It is flexibible enough to support very small and very large (multi-block) binary data.
+
+```sh
+type Index [Int]
````
Is this supposed to just be `type Index Int`, perhaps?

Edit:  I'm... not actually sure I sussed what this is for, so maybe I'm wrong and it's right.  You could write some short docs per type, using pound-prefixed lines (e.g. `# the docs`).

mikeal replied:
It’s correct. The index is a tuple of `[index, length]`. Ideally this would also enforce the length though, so I’m keen to hear any suggestions for improvements.

rvagg replied:
I'm not entirely clear on what "index" means in here. Just a list of parts and their lengths, and if so what is the first part of that tuple for? Couldn't it just be an array of lengths and the index is the zero-based index of that list?

If you want to enforce this as a strict tuple of two things then this is probably what you want:

```
type Index struct {
  index Int
  length Int
} representation tuple
```

But the duplication of "index" here could maybe be addressed to make it clearer what this is for?

mikeal replied:
factored out.

warpfork commented on schema-layer/data-structures/Data.md:
```diff
@@ -0,0 +1,62 @@
+`Data` is an advanced layout for representing binary data.
+
+It is flexibible enough to support very small and very large (multi-block) binary data.
```
flexible :)

mikeal replied:
It’s so flexible, it’s biblical: flexibible.

warpfork commented:
I have some mild observations at most, but overall assume you've thought about this and it represents progress... therefore, go ahead :)

Observations:

- The asymmetry between reading and writing -- the latter needing More Choosepower in order to actually be defined -- is... interesting.  I continue to be vaguely wary of this.  (I think eventually we're going to end up with a nasty word to describe "ADLs that require additional programmatic intervention to make writes be defined".)  But I won't go quite as far as to say we shouldn't explore it!  And I appreciate that it's clearly described.

- Similarly I'm just vaguely surprised this is as complex as it is.  There's... three...?... different advanced layouts described in the same file?  Which of these would I expect to start reading at, if I was going to do something with this, or implement something to match it?  Which of them would I choose when writing data as an application that has all three available as a library, and what heuristics would I use to make that choice as an application author?  The docs at the bottom of the file are good, but I think more granular docs on some of these types would help enormously.

- +1 to @rvagg 's comments about some of the field serial naming seeming to be different flavors of arbitrary.

- I'm not going to get too fussy about file paths at the moment, because I don't want to open the whole can of worms in a blocking way on this PR... but longer run: I suspect we'll want to make some sort of a directory where we can make something almost like a brief registry of various Advanced Layouts that are known to us.  And the ones authored by us as core people might eventually end up sharing that directory with at least a handful of especially reputable other contributed ones... so ideally, we should position things to look like just well-known library parts rather than blessed things sooner than later.  But again, not a blocking comment, we don't have to do the lifting on sorting that out right now.

mikeal commented:
> Similarly I'm just vaguely surprised this is as complex as it is. There's... three...?... different advanced layouts described in the same file? Which of these would I expect to start reading at, if I was going to do something with this, or implement something to match it? Which of them would I choose when writing data as an application that has all three available as a library, and what heuristics would I use to make that choice as an application author? The docs at the bottom of the file are good, but I think more granular docs on some of these types would help enormously.

Theoretically, this could just be one Advanced Layout. The reason I made it 3 is so that the `read()` methods would be attached directly to relevant data, rather than just implemented as branching logic in one large Advanced Layout at the root. 

It also makes the underlying data schemas and advanced layouts more usable independently. If I knew all my data had a low max chunking I could write a schema that just directly uses the ByteList. I couldn’t do that easily if all the read code was in a single place in the root Advanced Layout.

mikeal commented:
Ok, pushed a bunch of changes that are now ready for review.

I’m also going to need to go back and work on the implementation a bit to make sure none of these changes cause any problems. The current type generation code doesn’t have tuple representation either, so I’m going to need to implement that now ;)

rvagg commented on schema-layer/data-structures/data.md:
````diff
@@ -0,0 +1,65 @@
+
+```sh
+type Index struct {
+ offset Int
````
needs extra space indent, and the next line

mikeal commented:
Wow, ok, way simpler now. One thing lead to another and I was able to remove a bunch of things.

rvagg commented on schema-layer/data-structures/data.md:
```diff
@@ -0,0 +1,43 @@
+
+type BytesUnion union {
+  | Bytes bytes
+  | &Bytes link
```
what purpose does the link serve now? Since `ByteUnionList` is a list of links, so it could be made up of links to links to bytes? I think the only thing I can see it being useful for here is for the case of being able to use `raw` blocks--so a parent block just has `size:Int` and `bytes:CID` and following the CID gives you a raw block of bytes. Is that what this is for? How much do we need to care about that case?

mikeal replied:
if you look closely at the new spec, this is the *only* way to link to raw blocks and not embed bytes in the CBOR block. we definitely need this ;)

mikeal replied:
actually... maybe i’m wrong. since you can link to the kinded union you could just link to blocks.

i guess i hadn’t updated my mental model of this to the kinded union and was still thinking in terms of the keyed union. i’ll update.

rvagg commented:
Yeah, seems to make sense, aside from that one question I left about the utility of `&Bytes link` in the kinded union. Re naming, are they type names going to make for weird APIs our of codegen to code against? I don't suppose it matters too much if we decide to rename them later on, I'm just wondering about the case of having things like `Union` in a type name.

mikeal commented:
thinking we should probably rename this to “Byte Layout”

rvagg commented:
Here's the kinds of shapes I can see in the current spec:

Single blobs:

* `{size:100,bytes:<bytes>}`
* `{size:100,bytes:CID<raw>}`
* `{size:100,bytes:{lengths:[100],parts:[CID<raw>]}}`

Sharded:

* `{size:100,bytes:{lengths:[100,...],parts:[CID<raw>,...]}}`
* Recursive forms: `{size:100,bytes:{lengths:[100,...],parts:[{size:50,bytes:CID<raw>,...},CID<raw>...]}}`

The options for maybe-links are reduced because of the `&` in the `ByteUnionList` descriptor (as opposed to introducing `link` into a kinded union at some point). The only case where you could have a CBOR block with bytes inside of it and not use `raw` blocks is for a fully-inlined form, the first example above: `{size:100,bytes:<bytes>}`.

This doesn't seem to be a big problem, I just want to make sure it's intentional. One potential drawback is that you couldn't use this to compose a large file from many small files' data blocks if those small files are encoded with the first form--but I can't really think of great use-cases for that anyway.

mikeal commented:
As much as we do want to support full inlining for small byte representations, there are also big drawbacks in terms de-duplication when doing this, so i would actually prefer to limit the expression a little bit to only very small representations in which the entire layout is inline.

chafey commented:
Can we get this merged?

Stebalien commented:
One quick clarification: any reason not to make the `BytesUnion` the top-level object? If we want the size to avoid having to sum the lengths, I'd consider just including it in the `NestedByteList` definition.

That would make this fully recursive and it might help with finding content (i.e., if someone tells us to download some sub-dag of something we already have, we don't need to find the new `FlexibleByteLayout` root, we can just load the `BytesUnion` we already have.

mikeal commented:
@Stebalien good idea. we previously had a separate affordance for a single linked `raw` block which meant the struct was the only place to put `size`. Now that we only support inline bytes, which has a length once the block is decoded, and the nested byte list we can push the union up to the top level.

mikeal commented:
@chafey I’d like to finish at least one complete implementation before we merge the draft. Should be ready by the end of the month.

rvagg commented on data-structures/data.md:
````diff
@@ -0,0 +1,30 @@
+
+```sh
+type Lengths [Int]
+type ByteUnionList [&FlexibleByteLayout]
````
this `&` is a bit annoying, maybe we should be introducing `link` into a kinded union somewhere to make this linking optional, cause this can't be properly fully inlined where `parts` is used

rvagg replied:
I wonder if we could extend `FlexibleByteLayout` to be recursive:

```

type FlexibleByteLayout union {
  | Bytes bytes
  | NestedByteList map
  | FlexibleByteLayout link
} representation kinded
```

I _think_ that's allowed and reasonable since this is only really a type hint that can't be strongly enforced at the individual block level prior to loading the link

rvagg replied:
^ if we did that, then we could drop the `&` here and we get fully-inline as an option, I think?

rvagg replied:
```
type FlexibleByteLayout union {
  | Bytes bytes
  | NestedByteList map
  | &FlexibleByteLayout link
} representation kinded
```

I think might be the correct way of doing it

mikeal commented on data-structures/data.md:
```diff
@@ -0,0 +1,30 @@
+type NestedByteList struct {
+  lengths Lengths
+  parts ByteUnionList
+  algo optional String
```
As I’m implementing this I’m wondering what the full implications of this are.

First of all, if two algorithms produce the same graph, this will cause them not to match.

Second, this isn’t really enough. For instance, I just implemented a balanced layout algorithm but it also accepts a limit (max number of links per block) and that is necessary to reproduce a matching graph. But if we add it, either as another property or as an addition to the string, it’ll make a lot of matching graphs (any two limits that are both above the total number in the FBL) no longer match.

🤷‍♂️

Stebalien replied:
Personally, I'd store the algorithm outside. E.g., when adding a file, I'd record how the file was imported in the metadata.

mikeal replied:
this isn’t for the chunking algorithm, it’s for the layout algorithm used for balancing the tree.

Stebalien replied:
I'd store both in the file metadata.

mikeal replied:
ya, let’s cut it from this spec and put it in a container. efficient mutations, which is the whole point of storing this info, could be considered out of scope here. this is all a lot simpler if the only role of an FBL is to represent an arbitrarily large series of order bytes.

ribasushi commented:
My 2c coming from a different PoV: I am a bit concerned we are not taking maximum advantage of the constraints inherent to the specs to lock down as many places of ambiguity as possible.

As an example we are missing an opportunity to express that the amount of "elements" in these 2 lines absolutely must match: https://github.com/ipld/specs/pull/211/files#diff-db17d28d5379ecaf5b5229734c8fb9cbR10-R11

Without trying to turn this into a bikeshed, is there a reason we aren't doing something tighter like e.g. ( very pseudo-json ):
```
{
  size: payloadsize
  nested: [
    "raw bytes 1", 
    {
       size: payloadsize
       nested: [
         "raw bytes 2", 
         "raw bytes 3", 
       ], 
    },
    "raw bytes 4",
    {
       size: payloadsize
       nested: [
         "raw bytes 5", 
       ], 
    },
    "raw bytes 6",
  ]
}

mikeal commented:
The implementation of this latest revision is much cleaner than previous versions https://github.com/mikeal/js-fbl/blob/master/src/bare.js#L6-L31

warpfork commented:
I was painting some bikesheds in conversation with @Stebalien the other day and came to a very similar question as [riba's](https://github.com/ipld/specs/pull/211#issuecomment-600988043): why do we let the requirement that those two lists must be same-sized be punted to custom logic needed above this?

We can pretty easily get the schema to enforce the right thing by just making a single list of structs of two fields each:  something like this (the focus being on `BlobList [BlobListEntry]`; forgive the unrelated renamings, I'm just re-pasting something that had received other iterations as well):

```
type BlobList [BlobListEntry]
type BlobLength int
type BlobListEntry struct {
  len BlobLength
  blob &Blob
}
type ShardedBlob struct 
  total int
  parts BlobList
}
type PlainBlob bytes
type Blob union {
  | PlainBlob bytes
  | ShardedBlob map
} representation kinded
```

Are there reasons to prefer two separate lists?  Something performance, maybe?  But if so, this might deserve some discussion, because I bet any concerns there could be mitigated.

- If we're imagining separate lists would make it so one can just parse all the lengths and ignore the rest of the object with a sufficiently clever lazy parser, that might be interesting, but is also dubious (lazy parsers are really hard in practice, and either way needs documenting explicitly if it's a goal because otherwise it's just too esoteric to presume as a reason.
- If we're worried about serial size, I'm pretty sure popping a `representation tuple` clause onto the `type BlobListEntry` declaration means the encoded format would end up pretty darn well packed.
- If it's some other kind of performance concern... well, what is it?

mikeal commented:
There was a lot of iteration so I think the separate lists were due to a prior design constraint we may no longer have. Could do something like:

```sh
type NestedByte struct {
  length Int
  part &FlexibleByteList
} representation tuple

type NestedByteList [ NestedByte ]

type FlexibleByteLayout union {
  | Bytes bytes
  | NestedByteList list
} representation kinded
```

In fact, this matches exactly what gets passed around in the implementation which is then *converted* to the separate lists.

mikeal commented:
Ya, this cleans things up nicely, I’m going to update the spec. https://github.com/mikeal/js-fbl/commit/b33659a0024f6b7d82fe9d56cea893e379a6b187

rvagg commented:
This is a bit of a bikeshed but I just want to flag it as a non-blocking thing to consider: previously in this repo we've discussed recommending naming things that aren't strictly verifiable in a way that hints at that. `type NestedByte struct { length Int; part &FlexibleByteList; } representation tuple` is one such structure, where the length is describing something non-local and not easily verifiable, and it could easily lie. Something like `lengthHint` sends that signal to anyone consuming this data that it's not safe to rely on (you know people will rely on it and lead to the kinds of things @ribasushi has been showing with dag-pb).

rvagg commented:
@ribasushi made a point to me just now that `length` is something where we need to annotate this schema with additional language for implementers (i.e. a limitation of schemas but one we've accepted by saying these things should be in docs surrounding schemas). Something like:

> * The `length` field must be a positive integer whose value is greater than or equal to zero.
> * Implementations must be able to read and write `length` values representing byte lengths greater than 4GiB, i.e. values greater than fit into standard unsigned 32-bit integers.

Stebalien commented:
> Something like lengthHint sends that signal to anyone consuming this data that it's not safe to rely on (you know people will rely on it and lead to the kinds of things @ribasushi has been showing with dag-pb).

If it's there, people will use it. My concern is actually in the other direction: if we call it a "hint", people may not validate it because it doesn't _have_ to be correct.

mikeal commented:
> If it's there, people will use it. My concern is actually in the other direction: if we call it a "hint", people may not validate it because it doesn't have to be correct.

I can see that too.

Still, we have this issue here and elsewhere. We need to figure out what term we want to use and apply it everywhere.

What is a concise way to describe “unverified” or “insecure” information that we can attach to arbitrary nouns?

chafey commented:
Rather that storing the length, how about storing the offset from the beginning of the blob in each NestedByte struct?  This actually adds information (rather than duplicate - the length is already implicit in Bytes) and allows a client to calculate the length for a given NestedByte by subtracting its offset from the next NestedByte offset (well except for the last one, but the overall blob length should be in the metadata).  Another benefit of this approach is that if a client knows a given offset in the blob they want, they can use binary search to locate the NestedByte that contains the data they want rather than having to iterate through all of the NestedBytes.  This can be super helpful when storing a blob that has some kind of structure (e.g. ZIP, TAR) in FBL and being able to access internal chunks quickly (e.g. access a single file rapidly in that ZIP or TAR file by storing a directory of files + offsets in the metadata)

warpfork commented:
> Rather that storing the length, how about storing the offset from the beginning of the blob in each NestedByte struct?

Putting non-local information into a block makes it effectively unreusable / less convergent.  It also makes it harder to verify without dredging for more data... which... being in blocks not even directly linked from the block that you're verifying, gets odd indeed (and not necessarily even safe, if approached from the wrong angle).

> use binary search

I think you can already do that.  The `NestedByte.lengthHint` property is an accumulation of all the lengths below it, recursively.  So you just skim all of the lengthHints out of the `NestedByteList` and you can build the necessary index for that binary search from there.

chafey commented:
> Putting non-local information into a block makes it effectively unreusable / less convergent. It also makes it harder to verify without dredging for more data... which... being in blocks not even directly linked from the block that you're verifying, gets odd indeed (and not necessarily even safe, if approached from the wrong angle).

Yes it would - I should have started by suggesting that we fully separate the storage of Blob bits from the structure that stitches them together.  In this case, the Blob bits would be stored in a block without any enveloping structure at all.  This is a clear separation of concerns (blob bit storage vs the structure that stitches them together).  I have always viewed this as BLOB data + metadata use to connect BLOB data which has been chunked.  The current design combines these more than I am comfortable with.

Another thing to consider is making it such that an entire FLB could be stored in a single block (including the blob data).  This would allow efficient storage of very small blobs.  You would just need to make a union for the blob data which is either Bytes or a link to Bytes.

mikeal commented:
> Putting non-local information into a block makes it effectively unreusable / less convergent. It also makes it harder to verify without dredging for more data

We need to document this practice a bit.

Internalizing the merkle constraint that “links only go in one direction” into a design principal that context is *additive* as you link to child blocks and that each thing you link to should be as useful on its own as possible.

In other words, when children contain information that is dependent on their parent they are useless independently since there is no way to link to a parent.

mikeal commented:
Since we’re using a tuple representation the name of the field can be updated later without breaking the schema.

I think this is ready to merge now. Once there’s a consensus on that I’ll create a new issue on the naming so we can find a general term that works for everywhere we’re encountering this problem.

rvagg commented:
I'm fine with this as is. I'm bothered by the juxtaposition of "Flexible" and the `&` which makes it not as flexible as it could be (if the `&` was optional, kinded union style), but that's also a possible upgrade path for the future if it ends up being an annoyance in practice.

ribasushi commented on data-structures/data.md:
```diff
@@ -0,0 +1,28 @@
+  part &FlexibleByteList
+} representation tuple
+
+type NestedByteList [ NestedByte ]
```
The singular ( one element ) is clashing cognitively with "single byte". Perhaps  `BytesSegment` ?

ribasushi commented on data-structures/data.md:
````diff
@@ -0,0 +1,28 @@
+```sh
+type NestedByte struct {
+  lengthHint Int
+  part &FlexibleByteList
````
`FlexibleByteList` is not a defined type, something got lost in translation...

ribasushi commented on data-structures/data.md:
````diff
@@ -0,0 +1,28 @@
+
+```sh
+type NestedByte struct {
+  lengthHint Int
````
> What is a concise way to describe “unverified” or “insecure” information that we can attach to arbitrary nouns?

I am going to list a bunch of things back to back, **without** being happy with any of them: perhaps something will stand out...

```
length Int
lengthHint Int
unverifiedLength Int
insecureLength Int
uncertifiedLength Int
uncertLength Int
forgeableLength Int
untrustedLength Int
uncheckedLength Int
metaLength Int
unauditedLength Int
```

mikeal commented:
If someone writes incorrect lengths, what do we expect the behavior to be in an implementation? Understanding what the failure mode is would help a bit here.

Stebalien commented:
> If someone writes incorrect lengths, what do we expect the behavior to be in an implementation? Understanding what the failure mode is would help a bit here.

I'd like to declare exactly how/where we should return an error in the spec. If a the "part" part of a `NestedByte` doesn't match the given length _when we resolve the part_, we should return an error.

mikeal commented:
> I'd like to declare exactly how/where we should return an error in the spec. If a the "part" part of a NestedByte doesn't match the given length when we resolve the part, we should return an error.

Agree, but “when we resolve the part” is a bit too ambiguous. We “resolve” the entire graph when we replicate it but because replicators work on generic IPLD graphs they won’t be able to produce errors during replication because they won’t understand this data structure or when to apply this schema.

I think “when we **read** the part” is less ambiguous and doable, but as @ribasushi has mentioned before, this still leaves you open to an odd situation when reading a slice of the data that occurs *after* a miss-alignment of the size will not cause an error and will return the data slice based on the encoded size information rather than the raw part length. This is probably *fine* but if we’re going to document it as part of the spec it should be this detailed, because we’re effectively saying that “when reading a slice of the data, the encoded size information is relied upon and miss-alignments will only cause errors if the miss-aligned part is being read.”

Stebalien commented:
Fair enough. With respect to reading parts that _are_ correctly aligned when some aren't, my mental model here is that the "bad" sections are like corrupt disk sectors. You don't get an error until you actually try to read them.

ribasushi commented:
> mental model here is that the "bad" sections are like corrupt disk sectors. You don't get an error until you actually try to read them.

~This is not an accurate model. A misalignment somewhere in the DAG creates a "fork". If one seeks directly to the past-misalignment part, in the current model there is no way to determine the DAG is iffy - everything comes back correctly.~

~Only by happening to read through a "DAG-arm" that contains the disagreeing numbers, you can detect the problem. In other words - if you want to be sure, the current constructs require you to read and validate each individually sized unit from 0 to your desired offset.~

See my next comment below

Stebalien commented:
> This is not an accurate model. A misalignment somewhere in the DAG creates a "fork".

By "fork", I assume you mean that reading sequentially yields different results than seeking? That's a _bug_ in the current implementation, not something that _has_ to be true. We can validate offsets as we read and return errors when we encounter misaligned sections.

ribasushi commented:
While trying to explain/illustrate my concern I think I may have arrived at a better formalization and something approaching a solution to the issue I keep bringing up. I need to let this "bake" over the weekend, and will update the thread. For the time being assume I didn't make the above comment 🤔

chafey commented:
Is this design finalized?  I have a need for something like this and am hoping to use this if it can be finalized.

mikeal commented:
We’ve spent far too much time bike shedding on “naming things.”

I’ve pushed a few changes to naming and am closing the loop on bike shedding the names. I’ve also fixed the issue @riba found (reference name wasn’t migrated when we changed the name of the data structure).

I’m going to merge this in 24h unless there’s a strong objection. I’m also going to get the JS implementation in-line with the latest schema today.

warpfork commented:
One tiny ask: can we add a line of prose that states clearly that `FlexibleByteLayout` is the type we expect to encounter first" at the start of one of these documents?  I think it is strongly implied that this is the case, but it would make a reader more confident to see it stated, I think.

Otherwise my green checkmark from ages ago still applies; this all looks reasonable to me.

ribasushi commented:
This comment is made against this version of the proposed spec: https://github.com/ipld/specs/blob/1345e737ecdeff3ee182a7458c4ce0667a43770b/data-structures/data.md

> We’ve spent far too much time bike shedding on “naming things.”

I posit that the "bikeshedding on naming" is a symptom of an underlying lack of shared context. There are two parts to that:

1. Unlike most ADL's that are designed to be upgradeable/swappable, this particular ADL is **very** special as it encodes streams - the most fundamental building block of any other ADL. If we conjecture that one of the goals of IPLD is to foster "emergent convergence" within the merkle-world, this layout becomes part of the "uncoordinated convergence toolset", namely:
   1. near-universally-preferred byte-chunking set of parameters (in progress)
   2. near-universally-preferred intermediate "link" node representation (this spec)
   3. near-universally-preferred node-tree arangement (in progress together with i., likely a two stage shrub+trickle arrangement)

If the answer to 1. above is "no not really, this is definitely not the final version for streams, this is something we are playing with" - then this is fine, and my only request is that the spec reflects that ( `prescriptive`/`draft` ? ). Otherwise we are sending the wrong signal to folks like @chafey who [clearly want to start using this as the blessed way to write data](https://github.com/ipld/specs/pull/211#issuecomment-645456685)

If the answer however is "yeah, we need to nail this down and try not to change this anymore", we get to:

2. While the ADL spec in isolation seems straightforward, it is not clear how does it (or whether it can at all) reference already-existing raw blocks (which in turn is mandated by the answer to 1 being "yeah"). This was raised in https://github.com/ipld/specs/pull/211#issuecomment-573488097 but it is not clear where it left things. Specifically, if this is a "finalish version" we need to answer (here or in a consequent PR) the following:
   1. Is `FlexibleByteLayout` the actual entrypoint for a network-block (@warpfork's question  https://github.com/ipld/specs/pull/211#issuecomment-645535107)? If yes - we may want to explicitly state that and/or move that definition first for clarity
   2. Is linking to raw ( `0x55` ) CID's allowed at `part &FlexibleByteLayout`? I.e. does a raw node satisfy the `FlexibleByteLayout.Bytes` union, or does one **always** have to wrap a cbor around even the smallest set of bytes?

I think this covers everything that bothers me :)

rvagg commented:
> Is FlexibleByteLayout the actual entrypoint for a network-block (@warpfork's question #211 (comment))? If yes - we may want to explicitly state that and/or move that definition first for clarity

Yes, and this is something we've not got good answers to with schemas yet, we either need some kind of signal indicating top-level elements or a convention. For now I suggest we adopt the convention of putting top-level elements at the top of a schema doc and ordering things in reverse dependency order (the opposite of what you'd do in a C program).

> Is linking to raw ( 0x55 ) CID's allowed at part &FlexibleByteLayout? I.e. does a raw node satisfy the FlexibleByteLayout.Bytes union, or does one always have to wrap a cbor around even the smallest set of bytes?

This is supported by the current iteration just fine. There's no tie between Schemas and CBOR, it's just a matter of shape and type. `raw` blocks would be matched by `&Bytes`, `&Any` or the kinded union that appears in `FlexibleByteLayout`. Having `bytes` as one of the kinds means that if you hit a `raw` either as the top-level element (i.e. a `FlexibleByteLayout` decoder is handed a `CID<raw>`) or via `&FlexibleByteLayout`. The `bytes` kind would also be matched if it hits a CBOR or DAG-CBOR block that contains _only_ a byte string element, or a DAG-JSON block that contains _only_ a `{"/": { "bytes": String } }`. And you could mix and match all of those types (and more that we may add in the future!) with this.

So to do my exercise again to look at the kinds of shapes this Schema might allow:

```ipldsch
type FlexibleByteLayout union {
  | Bytes bytes
  | NestedByteList list
} representation kinded

type NestedByteList [ NestedByte ]

type NestedByte struct {
  length Int
  part &FlexibleByteLayout
} representation tuple
```

1. single `raw` or other format containing only one top-level `bytes`
2. `[ [ length, CID<x> ], ... ]` where `x` is either 1 or 2

And that's it I think.

Possible drawbacks:

* It allows very deep nesting if you keep on linking to lists within lists, e.g..: `[[ l, &[[ l, &[[ l, &[[ l, &[[ l, &[[ l, &[[ l, &[[ l, &[[ l, &[[ bytes ]] ]] ]] ]] ]] ]] ]] ]] ]] ]]` (where `&` is a separate, linked block). I'm not sure there's anything we can do about that, nor should we probably try, since it allows arbitrary concatenation of byte arrays.
* `length` is advisory only, but this continues to be a problem across our multi-byte data structures. I'd prefer the name to indicate it but will be satisfied with documentation about its advisory status.
* The only way to fully-inline is to do it as a top-level `bytes` (`raw` or a single top-level `bytes` type). But that's probably not a big deal, I can't think of a case where you'd want to fully-inline but also need to retain explicit splitting and supporting that is going to complicate implementation. Perhaps if you wanted to do this, you could embed `FlexibleByteLayout` under a top-level container that also includes metadata about the splits, but that would be something for whoever needs this functionality to do.
* You can't do partial inline, e.g. `[ [ length, <bytes> ], [ length, CID ] ]`. This might be useful if you wanted to form a stream while also minimising the number of blocks. I can imagine use-cases for this, but it's also something that could be added later if we decide it's a big deal (introducing a maybe-link between `NestedByte` and `FlexibleByteLayout`).

mikeal commented:
> One tiny ask: can we add a line of prose that states clearly that FlexibleByteLayout is the type we expect to encounter first" at the start of one of these documents? I think it is strongly implied that this is the case, but it would make a reader more confident to see it stated, I think.

How would we go about enforcing this?

The schema can be applied to any node. Whether or not the top level type is a discrete block is really a decision that libraries are going to have to make when they decide how they want to support it. It’s part of the implementation logic, which isn’t defined in the spec or the schema.

I don’t see a clear *problem* with applying this schema to a node that is within block rather than the root of the block, but maybe I’m missing something.

mikeal commented:
@rvagg pushed fixes for you comments.

mikeal commented:
> While the ADL spec in isolation seems straightforward, it is not clear how does it (or whether it can at all) reference already-existing raw blocks

You can use existing `raw` blocks, in fact, what you can’t do is embed binary values in the fully nested structure, you actually **MUST** use links. This isn’t all that clear because of how recursion works in the schema.

`NestedByte.part` is a link to the `FlexibleByteLayout` union. Since one of the union kinds is `bytes` that means `NestedByte.part` can be a link to a `raw` block.

The only opportunity you have to use a binary value without linking to a `raw` block is if the entire `FlexibleByteLayout` is a single binary value or block.

rvagg commented:
> How would we go about enforcing this?

Right, and I agree that it's not enforcement that we need here. It's more about clarity about the _expected entry point_ for the primary purpose of this schema. I expect we're going to find codegen and code interface reasons why we need more than just a convention (like the `advanced` keyword), but for now I think that making sure our documentation is clear about how these pieces fit together and their anticipated hierarchy.

As it is, this document has the same title as the top-level type, so it's now clear in that way as well as the presented order, so that's good enough for me for now. We can work on this challenge over time.

warpfork commented:
Thrilled that this is merged.

Quick recap on the discussion about "entry point types" in the schemas and how I feel about that:  Yeah, I think ya'll kinda covered all facets of it already.... :+1: 

- Docs for something like an ADL probably should make a habit of making it clear what the entry point type is... because that's defacto definitely something that exists _for a use case like ADLs_...
- but they should do it ***in prose***.
- I'll double down on this: I'm fairly convinced that codifying it would be wrong, and we can put that thought to bed.  Why?  Because "entry point type" is a concept that does _not_ exist for all use cases (as Mikeal also said in [this](https://github.com/ipld/specs/pull/211#issuecomment-646139457) comment; I totally agree) -- there are lots of times when I write one schema and want to use it for, say, a ton of different API endpoints in a web service; then there's lots of "entry point" types and we wouldn't want to write a new schema for each API endpoint!
- I'm also thumbs up on making it a conventional habit to do "inverse of C order" and put the "entry point" types towards the top.  That's how I usually write things, for whatever it's worth.

:ship:

chafey commented:
If the entrypoint matters it should be in prose.  It wasn't mentioned in the spec so I wasn't sure, I left thinking "it didn't matter, I can choose any".  It would be nice to have some examples in different languages of FBL implementations and of course a reader perhaps as part of a ADL Utility library.  If anyone has any ideas on what would be a good example use of this, please let me know as I'll probably work on this today.  

A few other cleanup items:
1) We should add a link to this spec from the main page 
2) We should rename the page from data.md to flexible-byte-layout.md

I'll submit a pr for these along with some prose on the entrypoint

warpfork commented:
(I think we're probably oscillating around the same understanding here, but just in case a little more discussion and alternative examples help...)

I think "I can choose any" is also still technically true, just defacto not so much in this case... because the "choose" is at the scale of "any ADL implementation that wants to be interoperable", so it's kinda just one big single incident of choose in practice.

Example in another domain that's an interesting contrast: in the [schema for Selectors](https://github.com/ipld/specs/blob/master/selectors/selectors.md#schema), I introduced a `SelectorEnvelope` type in addition to the `Selector` type, and recommended that one might want to use that when one is handling serializing and deserializing messages that don't have established existing context that can make it clear that the following data is a Selector.

But that sentence is in prose in the docs and is heavy in conditionals on purpose: I simultaneously consider it totally reasonable that if someone else is designing a larger schema that uses Selectors somewhere inside their messages, they might just use the `Selector` type rather than the `SelectorEnvelope` type.  Doing so would make perfect sense if they have additional surrounding structure like `type MyCoolProtocolMessage struct { otherInfo CoolTypes; selectorForBees Selector; selectorForBalloons Selector }`, for example, because there's already plenty of structural entropy there.
